diff --git a/drivers/irqchip/irq-bcm2835.c b/drivers/irqchip/irq-bcm2835.c
index a035e1ceacc0..7ffd4d339ce2 100644
--- a/drivers/irqchip/irq-bcm2835.c
+++ b/drivers/irqchip/irq-bcm2835.c
@@ -181,10 +181,7 @@ static void armctrl_ack_irq(struct irq_data *d)
 static struct irq_chip armctrl_chip = {
 	.name = "ARMCTRL-level",
 	.irq_mask = armctrl_mask_irq,
-	.irq_unmask = armctrl_unmask_irq,
-#ifdef CONFIG_ARM64
-	.irq_ack    = armctrl_ack_irq
-#endif
+	.irq_unmask = armctrl_unmask_irq
 };
 
 static int armctrl_xlate(struct irq_domain *d, struct device_node *ctrlr,
diff --git a/drivers/irqchip/irq-bcm2836.c b/drivers/irqchip/irq-bcm2836.c
index 6dccdf94f6f3..4764a19b0e84 100644
--- a/drivers/irqchip/irq-bcm2836.c
+++ b/drivers/irqchip/irq-bcm2836.c
@@ -196,7 +196,6 @@ __exception_irq_entry bcm2836_arm_irqchip_handle_irq(struct pt_regs *regs)
 		u32 ipi = ffs(mbox_val) - 1;
 
 		writel(1 << ipi, mailbox0);
-		dsb(sy);
 		handle_IPI(ipi, regs);
 #endif
 	} else if (stat) {
diff --git a/drivers/pinctrl/bcm/pinctrl-bcm2835.c b/drivers/pinctrl/bcm/pinctrl-bcm2835.c
index 81a66cba2ab0..592776f12125 100644
--- a/drivers/pinctrl/bcm/pinctrl-bcm2835.c
+++ b/drivers/pinctrl/bcm/pinctrl-bcm2835.c
@@ -106,7 +106,7 @@ struct bcm2835_pinctrl {
 	struct gpio_chip gpio_chip;
 	struct pinctrl_gpio_range gpio_range;
 
-	struct bcm2835_gpio_irqdata irq_data[BCM2835_NUM_IRQS];
+	struct bcm2835_gpio_irqdata irq_data[BCM2835_NUM_BANKS];
 	spinlock_t irq_lock[BCM2835_NUM_BANKS];
 };
 
@@ -394,6 +394,8 @@ static struct gpio_chip bcm2835_gpio_chip = {
 
 static int bcm2835_gpio_irq_handle_bank(struct bcm2835_pinctrl *pc,
 					unsigned int bank, u32 mask)
+// 1 line added below to apply patch
+static irqreturn_t bcm2835_gpio_irq_handler(int irq, void *dev_id)
 {
 	unsigned long events;
 	unsigned offset;
@@ -409,6 +411,12 @@ static int bcm2835_gpio_irq_handle_bank(struct bcm2835_pinctrl *pc,
 
 		generic_handle_irq(irq_linear_revmap(pc->irq_domain, gpio));
 	}
+	return events ? IRQ_HANDLED : IRQ_NONE;
+}
+
+static inline void __bcm2835_gpio_irq_config(struct bcm2835_pinctrl *pc,
+        unsigned reg, unsigned offset, bool enable)
+// 5 lines added above to apply patch
 
 	return (events != 0);
 }
